src/compiler/checker.ts

```diff
@@ -6616,7 +6616,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
             function isHomomorphicMappedTypeWithNonHomomorphicInstantiation(type: MappedType) {
                 return isMappedTypeWithKeyofConstraintDeclaration(type)
-                    && !(getModifiersTypeFromMappedType(type).flags & TypeFlags.TypeParameter);
+                    && !(getModifiersTypeFromMappedType(type).flags & TypeFlags.TypeParameter)
+                    && !!type.aliasSymbol;

```

Comment:
This shouldn't have anything to do with the alias symbol as by the time we're trying to print with this function, the alias symbol has already been discarded as an option for printing the mapped type. Instead, what we care about is if the type will be picked up as a homomorphic type variable (or not). Type parameters always are, so were a conservative initial choice - it's generally safe to wrap anything else which is still recognized as a homomorphic type variable up and present it as a type parameter, too - all you get is unnecessarily verbose printback sometimes. Instead of checking the alias symbol here (which seems pretty unrelated), you want to check, rather than if the modifiers type is still a type parameter, if an index on the modifiers type still produces a homomorphic type variable. In this case, an index of a mapped type just returns the inner mapped type's constraint type, which can then be unwrapped, so it is recognized as homomorphic. So, instead of the simple, conservative `!(getModifiersTypeFromMappedType(type).flags & TypeFlags.TypeParameter)` check, you want something to more closely mirror `getHomomorphicTypeVariable`'s conditions like
```ts
                let index: Type;
                return isMappedTypeWithKeyofConstraintDeclaration(type)
                    && (index = getIndexType(getModifiersTypeFromMappedType(type)), !(index.flags & TypeFlags.Index && (index as IndexType).type.flags & TypeFlags.TypeParameter)))
```

Reply:
I applied the suggested change but it reverted the change to the emit of this thing ([TS playground](https://www.typescriptlang.org/play?ts=5.1.0-dev.20230315#code/KYDwDg9gTgLgBAE2AYwDYEMrDjAnmbAeQFsBLGAHgBUAaOAaTlBmADsEBnOAa2FwgBmcKgD44AXjgAFUsm7U6AURBoArkgq9+Q2gxEiA3AChQkWIhQYsOfNimYYpdKilQIBWKWAcFDJiBZ2Li1BYTFJe09nChk5X3p9OAAyOBJyeMMTcGh4AVVWZEcIVkQIAGUIYmAYAAtSVgBzAH0AQULVaKp-QM44AG8jOCG4MDcwAC44DhgoeobjAF8RAAp0SaoASn7B4eRi6bgQSb64AG0pOHqePlDIx2dXd2BPb18AIlH3N5EAXUm7pwuMbPRyvXQfMbfc4-AxwBYSOCsVSoVBwdBcdCsXDGYZwLAwVRQEogRZAA)):
```ts
export declare type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export declare type PartialProperties<T, K extends keyof T> = Partial<Pick<T, K>> & Omit<T, K>;
export function doSomething_Actual<T extends {
    prop: string;
}>(a: T) {
    const x: { [P in keyof PartialProperties<T, "prop">]: PartialProperties<T, "prop">[P]; } = null as any;
    return x;
}
```

Does this one have to be rewritten? 🤔

Reply:
Probably not, since, in that case, I don't think `PartialProperties` is homomorphic, so the wrapping mapped type shouldn't be either. Right now we're using `isMappedTypeWithKeyofConstraintDeclaration` as a proxy for was-this-declared-as-homomorphic - we probably need to swap that out for something a bit more accurate, too. Potentially the same check we're doing on `type` but on `type.target` (and negated)?

Reply:
Indeed, it's not actually homomorphic - 
```ts
doSomething_Actual<{prop: string, propb: string} | {prop: string, propa: string}>(null as any);
```
ends up with type 
```ts
{
    prop?: string | undefined;
}
```
so we just need to tighten up that side of the check, too.

Reply:
This particular mapped type doesn't have a `.target` property - I might have misunderstood your recommendation for the proposed changes but for now, I've chosen to just avoid rewriting in case of `!type.target`. I think this is correct (it yields the expected result for this case and the rest of the test suite) but maybe I'm missing some further details.

Reply:
Actually, that's perfect - if the mapped type doesn't have a `.target`, then leaving it as-is is always correct - it's still it's declared form, uninstantiated.